server/src/main/java/io/crate/execution/dml/upsert/FromRawInsertSource.java

```diff
@@ -21,27 +21,148 @@
+        this.txnCtx = txnCtx;
+        this.nodeCtx = nodeCtx;
+        this.checks = new CheckConstraints<>(
+            txnCtx,
+            inputFactory,
+            new FromSourceRefResolver(table.partitionedByColumns(), indexName),
+            table
+        );
     }
 
     @Override
-    public Map<String, Object> generateSourceAndCheckConstraints(Object[] values) throws IOException {
-        return JsonXContent.JSON_XCONTENT.createParser(
-            NamedXContentRegistry.EMPTY,
-            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
-            new BytesArray(((String) values[0])).array()
-        ).map();
+    public Map<String, Object> generateSourceAndCheckConstraints(Object[] values) {
+        String src = (String) values[0];
+        Map<String, Object> parsed = Map.of();
+        try {
+            var parser = JsonXContent.JSON_XCONTENT.createParser(
+                NamedXContentRegistry.EMPTY,
+                DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                new BytesArray(src).array());
+            parsed = parser.map();
+            if (parser.getTokenLocation().columnNumber < src.length()) {
+                throw new IllegalArgumentException("failed to parse `" + src.substring(parser.getTokenLocation().columnNumber) + "`");
+            }
+        } catch (IOException e) {
+            if (e instanceof JsonParseException) {
+                throw new IllegalArgumentException("failed to parse");
+            }
+        }
+
+        Map<String, Object> validatedSource = validateRawSource(parsed);
+        mixinDefaults(validatedSource,
+                      // evaluateDefaultValues is invoked per row for expressions like now(), and gen_random_text_uuid().
+                      // Other deterministic expressions like abs(-10) should have been evaluated once and for all by the analyzer.
+                      evaluateDefaultValues(table.defaultExpressionColumns(), txnCtx, nodeCtx));
+
+        GeneratedColumns<Map<String, Object>> generatedColumns = new GeneratedColumns<>(
+            inputFactory,
+            txnCtx,
+            FromSourceRefResolver.WITHOUT_PARTITIONED_BY_REFS,
+            refLookUpCache.values(),
+            table.generatedColumns());
+
+        generatedColumns.setNextRow(validatedSource);
+        generatedColumns.validateValues(validatedSource);
+
+        for (var entry : generatedColumns.generatedToInject()) {
+            var reference = entry.getKey();
+            var value = entry.getValue().value();
+            var valueForInsert = reference
+                .valueType()
+                .valueForInsert(value);
+            var column = reference.column();
+            Maps.mergeInto(validatedSource, column.name(), column.path(), valueForInsert);
+        }
+
+        checks.validate(validatedSource);
+
+        return validatedSource;
+    }
+
+    private Map<String, Object> validateRawSource(Map<String, Object> rawSource) {
```

Comment:
would it be possible to do the validation in parsing in one pass using the parser/token API?
Would avoid creating a map twice.

Reply:
Please check a122c6a for updates, I tried to reuse but modify ES code as least as possible.

Reply:
You could also have a look at the `SourceParser` which does something similar. (But also supports skipping columns it doesn't need)

Reply:
I have checked `SourceParser` but it involves several nullchecks for dataTypes and requiredColumns. And if I were to have a version with the nullchecks removed, and add a step for validation, it will end up similar to https://github.com/crate/crate/commit/a122c6a2f1828b4bc3028c5f607c693fc8f1d043. 
Should I try to copy over the logic similar to `SourceParser`?